app/scripts/controllers/network/network-controller.js

```diff
@@ -476,4 +501,113 @@ export default class NetworkController extends EventEmitter {
     this._provider = provider;
     this._blockTracker = blockTracker;
   }
+
+  /**
+   * Network Configuration management functions
+   */
+
+  /**
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
+   *
+   * @param {NetworkConfiguration} networkConfiguration - The network configuration to add or, if rpcUrl matches an existing entry, to modify.
+   * @param {object} options
+   * @param {boolean} options.setActive - An option to set the newly added networkConfiguration as the active provider.
+   * @param {string} options.referrer - The site from which the call originated, or 'metamask' for internal calls - used for event metrics.
+   * @param {string} options.source - Where the upsertNetwork event originated (i.e. from a dapp or from the network form)- used for event metrics.
+   * @returns {string} id for the added or updated network configuration
+   */
+  upsertNetworkConfiguration(
+    { rpcUrl, chainId, ticker, nickname, rpcPrefs },
+    { setActive = false, referrer, source } = {},
```

Comment:
Should we make this mandatory, now that it includes `referrer` and `source`? Those seem pretty important to include

Reply:
Sure! Make it mandatory by throwing if it's not present?

Reply:
Mainly I was thinking we shouldn't have a default here on this line if they're non-optional:

```suggestion
    { setActive = false, referrer, source },
```

I guess we could use validation as well though

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/17421/commits/c7778e87ef17acda0e048306f1108a698164284c